src/ucs/datastruct/array.h

```diff
@@ -0,0 +1,224 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCS_ARRAY_H_
+#define UCS_ARRAY_H_
+
+#include <ucs/sys/compiler_def.h>
+#include <ucs/sys/preprocessor.h>
+
+BEGIN_C_DECLS
+
+
+/**
+ * Declare an array type.
+ * The array keeps track of its capacity and length (current number of elements)
+ * It can be either fixed-size or dynamically-growing.
+ *
+ * @param _name         Array definition name (needed for some array operations)
+ * @param _index_type   Type of array's index (e.g size_t, int, ...)
+ * @param _value_type   Type of array's values (could be anything)
+ */
+#define UCS_ARRAY_DECLARE_TYPE(_name, _index_type, _value_type) \
+    typedef struct { \
+        _value_type       *buffer; \
+        _index_type       length; \
+        _index_type       capacity; \
+    } ucs_array_t(_name); \
+
+
+/**
+ * Declare the function prototypes of an array
+ *
+ * @param _name          Array name
+ * @param _index_type    Type of array's index
+ * @param _value_type    Type of array's values
+ * @param _storage_class Storage class for array's functions (e.g 'static inline')
```

Comment:
maybe rename `_storage_class` to `_scope` (e.g. as in in KHASH - see `__KHASH_IMPL`)?